framework/src/base/NonlinearSystem.C

```diff
@@ -881,22 +896,47 @@ NonlinearSystem::enforceNodalConstraintsJacobian(SparseMatrix<Number> & jacobian
     std::vector<dof_id_type>& slave_nodes = nc->getSlaveNodeId();
     if (slave_nodes.size() > 0)
     {
-      Node & master_node = _mesh.node(nc->getMasterNodeId());
-      // reinit variables at the master node
-      _fe_problem.reinitNode(&master_node, tid);
-      _fe_problem.prepareAssembly(tid);
-
-      // go over slave nodes
-      for (std::vector<dof_id_type>::iterator it = slave_nodes.begin(); it != slave_nodes.end(); ++it)
+      std::vector<dof_id_type>& master_nodes = nc->getMasterNodeId();
+      // loop over the master nodes
+      for (std::vector<dof_id_type>::iterator it = master_nodes.begin(); it != master_nodes.end(); ++it)
       {
-        Node & slave_node = _mesh.node(*it);
-        // reinit variables on the slave node
-        _fe_problem.reinitNodeNeighbor(&slave_node, tid);
-        // compute Jacobian
-        nc->computeJacobian(jacobian);
-      }
+        Node & master_node = _mesh.node(*it);
+        //reinit variables at the master node
+        _fe_problem.reinitNode(&master_node, tid);
+        _fe_problem.prepareAssembly(tid);
+
+        // go over slave nodes
+        for (std::vector<dof_id_type>::iterator it = slave_nodes.begin(); it != slave_nodes.end(); ++it)
+        {
+          Node & slave_node = _mesh.node(*it);
+          // reinit variables on the slave node
+          _fe_problem.reinitNodeNeighbor(&slave_node, tid);
+          // compute Jacobian
+          nc->computeJacobian(jacobian);
+          //cache Jacobian entries
+          _fe_problem.cacheJacobian(0);
+        }
+      } //end master loop
     }
   }
+  #ifdef LIBMESH_HAVE_PETSC
+  //Necessary for speed
+  #if PETSC_VERSION_LESS_THAN(3,0,0)
```

Comment:
This `MatSetOption` code is copy/pasted from somewhere, right?  We should probably factor this out into a PetscSupport utility to avoid that.